packages/core/src/browser/preferences/preference-contribution.ts

```diff
@@ -53,6 +54,7 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
 
     protected readonly preferences: { [name: string]: any } = {};
     protected readonly combinedSchema: PreferenceDataSchema = { properties: {}, patternProperties: {} };
+    private remoteSchemas: IJSONSchema[] = [];
```

Comment:
What is special about `remoteSchema`? Why cannot be incorporated in `combinedSchema` and should be validated separately?

Reply:
>After reading code it does not seem to affect combined schemas, but just passed here to reuse `validate` function for launch folder providers?

You're right, but not only for launch folder provider but for any preference provider. Each time a launch configuration is added, renamed or removed we need new schema which restricts possible names of configuration available for compound sub-section.

> `PreferenceSchemaProvider` is responsible for providing a schema for settings files, validating it and used for getting default values of settings. It does not have anything to do with launch preferences.

User might have a global or workspace wide launch configuration placed in corresponding settings file, and a project might contain no launch file at all. That is why I treated launch configuration as any other property.

Reply:
> User might have a global or workspace wide launch configuration placed in corresponding settings file

Wow, did not know it, ok, will test how it works. So i should get content assist know in settings.json for `launch` property and registered debug configurations similar to VS Code?

Reply:
yes

Reply:
There is no content assist support for `launch` property and launch configurations in 'settings.json` as in VS Code. Looking at the code this property should be provided by https://github.com/akurinnoy/theia/blob/397f360422dae3dc590e58665fd526d04d7810c7/packages/debug/src/browser/debug-schema-updater.ts#L77-L88 not by remote schemas.

I'm still confused why we need `remoteSchemas` here? How do they affect settings schema and values?

Reply:
Thinking of content assist I meant this:

![launch-content-assist](https://user-images.githubusercontent.com/16220722/55388142-66583900-553b-11e9-8110-8ea46729445e.gif)

> I'm still confused why we need `remoteSchemas` here? How do they affect settings schema and values?

That's the point. Remote schema doesn't affect setting a values. It affects only validation. Remote schema only makes things easier allowing to replace a part of preference schema without calling `doSetSchema()`, thus allows to easily update validation function.

The main part of launch preference schema is set only once. Since then only remote part of schema become updated.
https://github.com/theia-ide/theia/pull/4743/files#diff-2dd1a479f035ae6ce8c1517929d4568dR70

Reply:
I will try again tomorrow. Tried the same today and did not get any content assist. The problematic part with it that these settings not respected by `DebugConfigurationWidget` and we don't have an implementation for `compounds` anywhere. We either have to implement it or keep content assist out of scope of this PR.